File: solutions/python/dispatch.py

import splunklib.client as client
import utils, sys, time

opts = utils.parse(sys.argv[1:], {}, ".splunkrc")
try: 

    if len(sys.argv) == 1:
        print "syntax: dispatch.py [list | run \"name of search\"]"
        sys.exit()
        
    params = {"host"        : opts.kwargs["host"],
              "port"        : opts.kwargs["port"],
              "username"    : opts.kwargs["username"],
              "password"    : opts.kwargs["password"],
              "owner"       : "-",
              "app"         : "-"}
    service = client.connect(**params)

    if sys.argv[1] == "list":
        print "List of all saved searches:"
        for s in service.saved_searches:
            print s.name
            print "  [%s]" % s.search
    elif sys.argv[1] == "run":
        ss = service.saved_searches[sys.argv[2]]
        print "Running " + ss.name
        job = ss.dispatch();
    
        sys.stdout.write("Job is running")
        sys.stdout.flush()
    
        while True:
            sys.stdout.write(".")
            sys.stdout.flush()
            time.sleep(1)
            job.refresh()
            if job.is_done():
                break    
        job.refresh()
        
        response = job.results()    
        print response

    else:
        raise Exception("unknown command " + sys.argv[1])
    
except Exception, err:
    sys.stderr.write('Error: %s\n' % str(err))

File: solutions/java/dispatch.java
import com.splunk.*; 
import java.util.*;
import java.io.*;

public class dispatch {

    public static void main(String[] args) {

		try {
			if(args.length < 1) {
				System.out.println("syntax: dispatch [list | run \"search name\"]");
				System.exit(0);
			}
			Command command = Command.splunk("info").parse(args);
	        Service service = Service.connect(command.opts);
			SavedSearchCollectionArgs ssca = new SavedSearchCollectionArgs();
			ssca.setOwner("-");
			ssca.setApp("-");
			ssca.setSearch(args[0]);
			SavedSearchCollection ssc = service.getSavedSearches(ssca);
			if(args[0].equalsIgnoreCase("run")) {
				if(!ssc.containsKey(args[1])) {
					throw new Exception("No such saved search: " + args[1]);
				}
				SavedSearch ss = ssc.get(args[1]);
				Job job = ss.dispatch();
				
	        	System.out.print("Job is running");
	        	while(!job.isDone()) {
	        		System.out.print(".");
	        		Thread.sleep(1000);
	        	}
	        	System.out.println("complete!");
		        InputStream response = job.getResults();
		        int chr = 0;
				while (chr != -1) System.out.write(chr = response.read());
			}
			else if(args[0].equalsIgnoreCase("list")) 
				for (SavedSearch s: ssc.values()) 
					System.out.println(s.getName()); 
			else 
				throw new Exception("unknown command " + args[0]);
		} 
    	catch (Exception e)  {
    		if(e.getMessage() != null) {
    			System.err.println("Error message: " + e.getMessage());
    		} else {
    			System.err.println("Error type: " + e.getClass());
    		}
		}
	}
}